maven-plugin/src/main/java/org/sonatype/nexus/maven/staging/StagingDeployMojo.java

```diff
@@ -111,6 +156,121 @@ private void doExecute() throws MojoFailureException, MojoExecutionException {
+          "Staging failed: staging directory points to an existing file but is not a directory or is not writable!");
+    }
+    else if (!target.exists()) {
+      target.mkdirs();
+    }
+
+    try {
+      String url = target.getCanonicalFile().toURI().toURL().toExternalForm();
+      return artifactRepositoryFactory.createDeploymentArtifactRepository(id, url, artifactRepositoryLayout, true);
+    }
+    catch (IOException e) {
+      throw new MojoExecutionException(
+          "Staging failed: staging directory path cannot be converted to canonical one!", e);
+    }
+  }
+
+  private void installLocally(
+      final File index,
+      final Artifact artifact,
+      final ArtifactRepository artifactRepository,
+      final String tag) throws ArtifactInstallationException
+  {
+    try {
+      //lock to make thread safe code since working with IO
+      readWriteLock.lock();
+
+      try {
+        artifactInstaller.install(artifact.getFile(), artifact, artifactRepository);
+        attachToIndex(index, artifact, tag, artifactRepository);
+      }
+      catch (IOException e) {
+        getLog().error("error accessing files for local installation: ", e);
+        throw new ArtifactInstallationException(e);
+      }
+      catch (ArtifactInstallationException e) {
+        getLog().error("error installing artifact: ", e);
+        throw e;
+      }
+    }
+    finally {
+      readWriteLock.unlock();
+    }
+  }
+
+  private void attachToIndex(
+      final File index,
+      final Artifact artifact,
+      final String tag,
+      final ArtifactRepository artifactRepository) throws IOException
+  {
+
+    Optional<String> maybePluginPrefix = artifact.getMetadataList()
+        .stream()
+        .filter(metadata -> metadata instanceof GroupRepositoryMetadata)
+        .map(metadata -> ((GroupRepositoryMetadata) metadata).getMetadata().getPlugins())
+        .filter(plugins -> !plugins.isEmpty())
+        .flatMap(List::stream)
+        .map(Plugin::getPrefix)
+        .findFirst();
+
+    Optional<String> maybePomFileName = artifact.getMetadataList()
+        .stream()
+        .filter(metadata -> metadata instanceof ProjectArtifactMetadata)
+        .map(projectMetadata -> projectMetadata.getLocalFilename(artifactRepository))
+        .findFirst();
+
+    Optional<String> maybeClassifier = Optional.ofNullable(artifact.getClassifier());
+
+    ArtifactInfo artifactInfo = new ArtifactInfo();
+    artifactInfo.setGroup(artifact.getGroupId());
+    artifactInfo.setArtifactId(artifact.getArtifactId());
+    artifactInfo.setVersion(artifact.getVersion());
+    artifactInfo.setTag(tag);
+    artifactInfo.setClassifier(maybeClassifier.orElse("n/a"));
+    artifactInfo.setPackaging(artifact.getType());
+    artifactInfo.setExtension(artifact.getArtifactHandler().getExtension());
+    artifactInfo.setPomFileName(maybePomFileName.orElse("n/a"));
+    artifactInfo.setPluginPrefix(maybePluginPrefix.orElse("n/a"));
```

Comment:
probably should define a constant for `n/a`

Reply:
done , thx for pointing it